internal/kms/kms/gcp/gcp.go

```diff
@@ -8,328 +8,63 @@ SPDX-License-Identifier: AGPL-3.0-only
-
-  - cloudkms.cryptoKeyVersions.useToDecrypt
-
-  - cloudkms.cryptoKeyVersions.useToEncrypt
-
-  - cloudkms.importJobs.create
-
-  - cloudkms.importJobs.get
-
-  - cloudkms.importJobs.useToImport
+  - cloudkms.cryptoKeys.get
+  - cloudkms.cryptoKeys.encrypt
+  - cloudkms.cryptoKeys.decrypt
 */
 package gcp
 
 import (
 	"context"
-	"crypto/rand"
-	"crypto/rsa"
-	"crypto/sha1"
-	"errors"
 	"fmt"
 	"io"
-	"strings"
-	"time"
 
-	kms "cloud.google.com/go/kms/apiv1"
-	"cloud.google.com/go/kms/apiv1/kmspb"
-	"github.com/edgelesssys/constellation/v2/internal/kms/config"
 	kmsInterface "github.com/edgelesssys/constellation/v2/internal/kms/kms"
-	"github.com/edgelesssys/constellation/v2/internal/kms/kms/util"
+	"github.com/edgelesssys/constellation/v2/internal/kms/kms/internal"
 	"github.com/edgelesssys/constellation/v2/internal/kms/storage"
-	"github.com/googleapis/gax-go/v2"
-	"google.golang.org/api/option"
-	"google.golang.org/grpc/codes"
-	"google.golang.org/grpc/status"
+	"github.com/edgelesssys/constellation/v2/internal/kms/uri"
+	gcpckms "github.com/hashicorp/go-kms-wrapping/wrappers/gcpckms/v2"
 )
 
-type clientAPI interface {
-	io.Closer
-	CreateCryptoKey(context.Context, *kmspb.CreateCryptoKeyRequest, ...gax.CallOption) (*kmspb.CryptoKey, error)
-	CreateImportJob(context.Context, *kmspb.CreateImportJobRequest, ...gax.CallOption) (*kmspb.ImportJob, error)
-	Decrypt(context.Context, *kmspb.DecryptRequest, ...gax.CallOption) (*kmspb.DecryptResponse, error)
-	Encrypt(context.Context, *kmspb.EncryptRequest, ...gax.CallOption) (*kmspb.EncryptResponse, error)
-	GetKeyRing(context.Context, *kmspb.GetKeyRingRequest, ...gax.CallOption) (*kmspb.KeyRing, error)
-	ImportCryptoKeyVersion(context.Context, *kmspb.ImportCryptoKeyVersionRequest, ...gax.CallOption) (*kmspb.CryptoKeyVersion, error)
-	UpdateCryptoKeyPrimaryVersion(context.Context, *kmspb.UpdateCryptoKeyPrimaryVersionRequest, ...gax.CallOption) (*kmspb.CryptoKey, error)
-	GetImportJob(context.Context, *kmspb.GetImportJobRequest, ...gax.CallOption) (*kmspb.ImportJob, error)
-}
-
 // KMSClient implements the CloudKMS interface for Google Cloud Platform.
 type KMSClient struct {
-	projectID        string
-	locationID       string
-	keyRingID        string
-	newClient        func(ctx context.Context, opts ...option.ClientOption) (clientAPI, error)
-	waitBackoffLimit int
-	storage          kmsInterface.Storage
-	protectionLevel  kmspb.ProtectionLevel
-	kekID            string
-	opts             []gax.CallOption
+	kms    *internal.KMSClient
+	client io.Closer
 }
 
 // New initializes a KMS client for Google Cloud Platform.
-func New(ctx context.Context, projectID, locationID, keyRingID string, store kmsInterface.Storage, protectionLvl kmspb.ProtectionLevel, kekID string, opts ...gax.CallOption) (*KMSClient, error) {
+func New(ctx context.Context, store kmsInterface.Storage, cfg uri.GCPConfig) (*KMSClient, error) {
 	if store == nil {
 		store = storage.NewMemMapStorage()
 	}
```

Comment:
Maybe this would be better put into the "no-store" case of setup.go? And then rename "no-store" to "local-store" or "mem-store". Would make the behaviour a bit more explicit imo.

Reply:
no-store != in-memory storage

The first case is just used as a placeholder, since cKMS does not use any kind of storage, the later is used to simply store keys in a map (everything is lost on restart). This is only used for unit testing, so I guess we could remove it from the default initialization functions and simply throw an error if no store is set

Reply:
Yes. But because of this check "no-store" will create a MemMapStorage in `setup.go:KMS(..., "no-store", ..)`, which seems like an odd behavior to me. I forgot that we need `no-store` for cKMS. Throwing an error seems like a good solution to me.